hazelcast/src/main/java/com/hazelcast/client/impl/ClientEndpointStatisticsManagerImpl.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.impl;
+
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+import static com.hazelcast.internal.util.MapUtil.createHashMap;
+
+public class ClientEndpointStatisticsManagerImpl implements ClientEndpointStatisticsManager {
+
+    private static final Function<String, ClientEndpointStatistics> STATS_CREATOR = key -> new ClientEndpointStatistics();
+
+    private final ConcurrentHashMap<String, ClientEndpointStatistics> stats = new ConcurrentHashMap<>();
+    private volatile long lastResetTime;
+
+    @Override
+    public void onEndpointAuthenticated(ClientEndpoint endpoint) {
+        ClientEndpointStatistics stats = getStats(endpoint);
+        stats.incrementConnectionsOpenedCount();
+        stats.addClientVersion(endpoint.getClientVersion());
+    }
+
+    @Override
+    public void onEndpointDestroyed(ClientEndpoint endpoint) {
+        if (!endpoint.isAuthenticated()) {
+            return;
+        }
+
+        ClientEndpointStatistics endpointStats = getStats(endpoint);
+        endpointStats.incrementConnectionsClosedCount();
+        long uptime = getUptime(System.currentTimeMillis(), endpoint);
+        if (uptime > 0) {
+            endpointStats.incrementTotalConnectionDuration(uptime);
+        }
+        endpointStats.addClientVersion(endpoint.getClientVersion());
+    }
+
+    @Override
+    public Map<String, ClientEndpointStatisticsSnapshot> getSnapshotsAndReset(Collection<ClientEndpoint> activeEndpoints) {
+        long now = System.currentTimeMillis();
+        for (ClientEndpoint endpoint : activeEndpoints) {
+            if (!endpoint.isAuthenticated()) {
+                continue;
+            }
+            ClientEndpointStatistics stat = getStats(endpoint);
+            long uptime = getUptime(now, endpoint);
+            if (uptime > 0) {
+                stat.incrementTotalConnectionDuration(uptime);
```

Comment:
Does it make sense to keep it in-memory and increment? We could also set it to `now - creationTime` and that would have the same result (unless I overlook something).

Reply:
If we were reporting all-time total connection duration, `now-creationTime` would work. But, we planned to send stats in 24-hour intervals and reset them afterward. Do you think that we should store and report all-time records instead?

I think this way, it is much easier to query, you just sum up values on a certain time interval. If we were reporting all-time values, queries have to sum up values reported recently by the unique nodes.